coralproject/talk
# [CORL-2316]: Add moderator messaging option

## What does this PR do?

This PR adds in the option to send a commenter a message if you are a moderator without changing their status (i.e. without a warning attached). This is for https://vmproduct.atlassian.net/browse/CORL-2316. 

Mod messaging is very similar to sending a warning message except that it doesn't change the user's status in any way.

## What changes to the GraphQL/Database Schema does this PR introduce?

To the GraphQL schema, these changes add `ModMessageStatus` and `ModMessageStatusHistory`, as well as adds these under `modMessage` to the `UserStatus`. These include the current status of moderation messages (is there currently an active mod message?) as well as the history of past moderation messages and their acknowledgment. In addition, these changes add `ModMessageUserInput`, `ModMessageUserPayload`, `AcknowledgeModMessageInput`, and `AcknowledgeModMessagePayload`. They also add `modMessageUser` and `acknowledgeModMessage`.

To the database schema, these changes add `ModMessageStatus` and `ModMessageStatusHistory`, as well as adds these under `modMessage` to the `UserStatus`. These changes add `modMessageUser` and `acknowledgeModMessage`. They also add `ConsolidatedModMessageStatus` and `consolidateUserModMessageStatus` to grab the message from the latest active moderation message.

## How do I test this PR?

 To test this, you can create at least one user that is a commenter. Then
 1. Logged in as admin/moderator, navigate to `Community` in the admin. 
 1. Find the user to whom you want to send a moderation message in the user table and click their status. Select `Message` at the bottom of the options.
 1. You should see a modal open with an area for entering a message to the user. Enter a message and click `Message User`. You should see a success modal confirming that a message was sent to the user.
 1. Now log in as the user to whom you sent the message. At the top of the stream, you should see the moderation message that you sent displayed with an option to acknowledge the message. See that you are still able to comment on the story.
 
 Other things to test with this flow include:
 - As the admin/moderator, visit the user's account history throughout the process and confirm that the message and message acknowledgment history is visible there.
 - If multiple messages have been sent by the moderator and unacknowledged, only the latest active moderation message is displayed at the top of the stream for acknowledgment.
 - A user sees both a warning and a moderation message to acknowledge if they have both.
 
## How do we deploy this PR?

I don't think there's anything that needs to be done on deploy.
